src/continuous_integration.md

```diff
@@ -0,0 +1,36 @@
+# Continuous Integration
+
+We recommend that you establish a CI build for your Relm4 app. This guide describes how to do it, and the caveats you must observe to make the build work.
+
+## GitHub Actions
+
+Starting with Relm 0.6.1, you can set up a CI build for your app on GitHub Actions.
+
+If your project builds with Cargo in a reasonably self-contained way (e.g. using `build.rs` to do packaging rather than external shell scripts), you can do this with an almost entirely standard Cargo build workflow, as shown below.
+
+**Caveats:**
+
+- You must `apt-get install` the relevant system libraries for the parts of GTK or Adwaita that your Relm4 app uses.
+- You must synchronise the GNOME version your app uses with the GNOME version available on the GitHub Actions runner. (For example, on the standard Ubuntu 22.04 runner, you must use GNOME 42.) Set this in Cargo.toml with `relm4 = { features = ["gnome_<version>"] }`.
+
+```yaml
+name: Rust
+
+on:
+  push:
+    branches: [ "main" ]
+  pull_request:
+    branches: [ "main" ]
+
+jobs:
+  build:
+    runs-on: ubuntu-22.04
+    steps:
+    - uses: actions/checkout@v3
+    - run: sudo apt-get update
+    - run: sudo apt-get install -y libgtk-4-dev libadwaita-1-dev
```

Comment:
It's probably better to just pull the gtk-rs image which already has everything installed: https://github.com/gtk-rs/gtk4-rs/pkgs/container/gtk4-rs%2Fgtk4

Also, it might be worth noting that `relm4-template` already comes with CI configurations for both GitHub and GitLab.

Reply:
Thanks - I have tested the following config which uses the `gtk-rs` Docker image, and I get a passing build:

```yml
jobs:
  build:
    runs-on: ubuntu-latest
    container:
      image: ghcr.io/gtk-rs/gtk4-rs/gtk4:latest
    steps:
    - uses: actions/checkout@v3
    - uses: actions-rs/toolchain@v1
      with:
        profile: minimal
        toolchain: stable
    - name: Build
      run: cargo build
    - name: Test
      run: cargo test
```

**But there is one problem** with the gtk4-rs image at the moment - a lack of version tags:

- `gtk4-rs` only has a `latest` tag (i.e. I can't pin to the GTK4 minor version that I'm using)
- it depends on the `gtk4-rs-core:latest` image (so that's not pinned)
- the `gtk4-rs-core` image in turn depends on `fedora:latest` - which (a) is not version pinned and (b) is maybe not the best distro to use in a CI environment because it is fast-moving

So I think some version pins need to be added within the gtk4-rs image before it can be totally safe to rely on for reproducible CI build behaviour

Reply:
The problem with the current approach is ironically the opposite of the problem with the gtk-rs image. With Ubuntu 22.04 you are basically pinned to a very old version of GTK and can't use more recent feature flags. The only chance to upgrade comes every two years when a new Ubuntu LTS becomes available as GitHub runner.

Therefore, I think the gtk-rs image is the better choice in general. Unless you explicitly enable feature flags in your Cargo.toml, your code should never touch features of newer GTK versions so it doesn't matter if you for example have GTK 4.8 or 4.12 on the CI system to detect problems. The only use-case where your approach is better is when you want to explicitly target Ubuntu 22.04 and want to be 100% sure it works there (which to be fair helped uncover an issue with Relm4's feature flags), but this is not what everyone want's to do by default.